src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2742,4 +2743,163 @@ namespace winrt::TerminalApp::implementation
+    // Return Value:
+    // - <none>
+    void TerminalPage::_EndPreviewColorScheme()
+    {
+        // Get the focused control
+        if (const auto activeTab{ _GetFocusedTabImpl() })
+        {
+            if (auto activeControl = activeTab->GetActiveTerminalControl())
+            {
+                // Get the runtime settings of the focused control
+                auto controlSettings = activeControl.Settings().as<TerminalSettings>();
+
+                // Get the control's root settings, the ones that we actually
+                // assigned to it.
+                auto parentSettings = controlSettings.GetParent();
+                while (parentSettings.GetParent() != nullptr)
+                {
+                    parentSettings = parentSettings.GetParent();
+                }
+
+                // If the root settings are the same as the ones we stashed,
+                // then reset the parent of the runtime settings to the stashed
+                // settings. This condition might be false if the settings
+                // hot-reloaded while the palette was open. In that case, we
+                // don't want to reset the settings to what they were _before_
+                // the hot-reload.
+                if (_originalSettings == parentSettings)
+                {
+                    // Set the original settings as the parent of the control's settings
+                    activeControl.Settings().as<TerminalSettings>().SetParent(_originalSettings);
+                }
+
+                activeControl.UpdateSettings();
+            }
+        }
+        _originalSettings = nullptr;
+    }
+
+    // Method Description:
+    // - Preview handler for the SetColorScheme action.
+    // - This method will stash the settings of the current control in
+    //   _originalSettings. Then it will create a new TerminalSettings object
+    //   with only the properties from the ColorScheme set. It'll _insert_ a
+    //   TerminalSettings between the control's root settings (built from
+    //   CascadiaSettings) and the control's runtime settings. That'll cause the
+    //   control to use _that_ table as the active color scheme.
+    // Arguments:
+    // - args: The SetColorScheme args with the name of the scheme to use.
+    // Return Value:
+    // - <none>
+    void TerminalPage::_PreviewColorScheme(const Settings::Model::SetColorSchemeArgs& args)
+    {
+        // Get the focused control
+        if (const auto activeTab{ _GetFocusedTabImpl() })
+        {
+            if (auto activeControl = activeTab->GetActiveTerminalControl())
+            {
+                if (const auto scheme = _settings.GlobalSettings().ColorSchemes().TryLookup(args.SchemeName()))
+                {
+                    // Get the settings of the focused control and stash them
+                    auto controlSettings = activeControl.Settings().as<TerminalSettings>();
+                    // Make sure to recurse up to the root - if you're doing
```

Comment:
I guess I'm happily relying on the fact that there's only ever at most three layers here, and there's definitely always two.
```
root settings (from CascadiaSettings)
^
|
previewed settings (may or may not exist)
^
|
control's runtime settings
```

We're mostly going to the root here just to make sure that we get the original settings, so that _later_, in EndPreview, we can reset the parent of the control to the original root settings. If there were ever more layers here, this would start to break down, sure.

Reply:
I might be able to write a test for this that would catch if we broke this assumption

Reply:
Okay, I added tests, yay!